packages/itwin/tree-widget/README.md

```diff
@@ -161,66 +152,151 @@ This package provides building blocks for custom models tree:
         treeRenderer={(props) => <CustomModelsTreeRenderer {...props} {...rendererProps} />}
       />
     </TreeWithHeader>
   );
 }
-
-type VisibilityTreeRendererProps = ComponentPropsWithoutRef<typeof VisibilityTreeRenderer>;
-type CustomModelsTreeRendererProps = Parameters<ComponentPropsWithoutRef<typeof VisibilityTree>["treeRenderer"]>[0];
-
-function CustomModelsTreeRenderer(props: CustomModelsTreeRendererProps) {
-  const getLabel = useCallback<Required<VisibilityTreeRendererProps>["getLabel"]>(
-    (node) => {
-      const originalLabel = props.getLabel(node);
-      return <>Custom node - {originalLabel}</>;
-    },
-    [props.getLabel],
-  );
-  return <VisibilityTreeRenderer {...props} getLabel={getLabel} getSublabel={getSublabel} />;
-}
 ```
 
+  <!-- END EXTRACTION -->
+
 #### Displaying a subset of the tree
 
 Models tree allows displaying a subset of all nodes by providing a `getFilteredPaths` function, which receives a `createInstanceKeyPaths` function for creating hierarchy node paths from instance keys or an instance label and returns a list of hierarchy node paths targeting some nodes. When these paths are provided, the displayed hierarchy consists only of the targeted nodes, their ancestors, and their children. Example implementation of `getFilteredPaths`:
 
+ <!-- [[include: [Presentation.Tree-widget.Get-filtered-paths-component-example, Presentation.Tree-widget.Get-filtered-paths-example], tsx]] -->
+ <!-- BEGIN EXTRACTION -->
+
 ```tsx
-const getFilteredPaths = async ({ createInstanceKeyPaths }) => {
-  return createInstanceKeyPaths({
-    // list of instance keys representing nodes that should be displayed in the hierarchy
-    keys: myInstanceKeys,
-    // instead of providing instance keys, a label can be provided to display all nodes that contain it
-    // label: "MyLabel"
-  });
-};
+type useModelsTreeProps = Parameters<typeof useModelsTree>[0];
+
+interface CustomCategoriesTreeProps {
+  getFilteredPaths: useModelsTreeProps["getFilteredPaths"];
+  viewport: Viewport;
+  selectionStorage: SelectionStorage;
+  imodel: IModelConnection;
+}
+
+function CustomModelsTreeComponent({ getFilteredPaths, viewport, selectionStorage, imodel }: CustomCategoriesTreeProps) {
+  const { modelsTreeProps, rendererProps } = useModelsTree({ activeView: viewport, getFilteredPaths });
+
+  return (
+    <VisibilityTree
+      {...modelsTreeProps}
+      getSchemaContext={getSchemaContext}
+      selectionStorage={selectionStorage}
+      imodel={imodel}
+      treeRenderer={(props) => <VisibilityTreeRenderer {...props} {...rendererProps} />}
+    />
+  );
+}
+
+<CustomModelsTreeComponent
+  selectionStorage={unifiedSelectionStorage}
+  imodel={imodel.imodel}
+  viewport={testViewport}
+  getFilteredPaths={async ({ createInstanceKeyPaths }) => {
+    return createInstanceKeyPaths({
+      // list of instance keys representing nodes that should be displayed in the hierarchy
+      targetItems: [imodel.physicalModel],
+    });
+  }}
```

Comment:
A lot of stuff for a single-function example. Can we only provide that one function like we had previously?

Reply:
Personally I think having an example of how to insert the filtering function into the custom tree as the function type is exported and needs to be redefined through parameter.

Reply:
Makes sense, and with the removal of the following part of the section there's no more duplication. Just make sure the code example compiles. Now it's combined from two pieces and the second piece is kind of hanging in no-mans land

Reply:
The example compiles, but the user who would be using the snippet would have to change the targetItems to their own as I could not figure out a way to provide universally working keys.

Reply:
Here's what I see in the README now:

```tsx
type useModelsTreeProps = Parameters<typeof useModelsTree>[0];

interface CustomModelsTreeProps {
  getFilteredPaths: useModelsTreeProps["getFilteredPaths"];
  viewport: Viewport;
  selectionStorage: SelectionStorage;
  imodel: IModelConnection;
}

function CustomModelsTreeComponent({ getFilteredPaths, viewport, selectionStorage, imodel }: CustomModelsTreeProps) {
  const { modelsTreeProps, rendererProps } = useModelsTree({ activeView: viewport, getFilteredPaths });

  return (
    <VisibilityTree
      {...modelsTreeProps}
      getSchemaContext={getSchemaContext}
      selectionStorage={selectionStorage}
      imodel={imodel}
      treeRenderer={(props) => <VisibilityTreeRenderer {...props} {...rendererProps} />}
    />
  );
}

<CustomModelsTreeComponent
  selectionStorage={unifiedSelectionStorage}
  imodel={imodel.imodel}
  viewport={testViewport}
  getFilteredPaths={async ({ createInstanceKeyPaths }) => {
    return createInstanceKeyPaths({
      // list of instance keys representing nodes that should be displayed in the hierarchy
      targetItems: [imodel.physicalModel],
    });
  }}
/>,
```

The `<CustomModelsTreeComponent ... />` is definitely misplaced and wouldn't compile if we take the example as is. On the other hand, do we even need that last part? Maybe `getFilteredPaths` function could be defined inside the `CustomModelsTreeComponent`? BTW, currently the function is not wrapped with `useCallback`, which, I believe, may be a bug (adding @saskliutas to confirm).